File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        sp: ListNode = head
        fp: ListNode = head.next

        while fp != None and fp.next != None:
            sp = sp.next
            fp = fp.next.next
        
        # Reverse 
        second = sp.next
        prev = sp.next = None

        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp
            
        second = prev
        first = head
        
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first = tmp1
            second = tmp2

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <unordered_map>
class Solution {
public:
    bool hasCycle(ListNode* head) {
        unordered_map<int,int> myMap;

        while (head != nullptr) {
            if (myMap.find(head->val) != myMap.end()) return true;
            else {
                myMap[head->val] = head->val;
                head = head->next;
            }
        }

        return false;
        // Check if value of head is in myMap
        // If yes, return true 
        // Else store the value of head in myMap
        // go to next node
    }
};
